### PR TITLE
Introduce the concept of an organisation

### PR DESCRIPTION
**Pull request by ale-rt**

### CODE REVIEW

File: src/euphorie/client/browser/organisation.py [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+    @property
+    @memoize
+    def organisation_title(self):
+        """The title of the organisation bound to this account (if it exists)
+        ot the account login name
+        """
+        organisation = self.organisation
+        if organisation is not None:
+            return organisation.title
+        return self.webhelpers.get_current_account().loginname
```
[the remark this answers is not in the document]
Comment (ale-rt) [reply src/euphorie/client/browser/organisation.py:113]: When the PR was created `first_name` was not even a thing.
I think proto did move forward in the last months and this PR did not catch up.

File: src/euphorie/client/browser/templates/organisation.pt [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                          </h4>
+                          <p class="field role" i18n:translate="">
+                            Owner
+                          </p>
+                        </li>
+                        <li class="row"
+                            tal:repeat="membership python:view.get_memberships(organisation)"
+                        >
+                          <h4 class="user field">
+                            <strong class="name">${membership/Account/loginname}</strong>
```
[the remark this answers is not in the document]
Comment (ale-rt) [reply src/euphorie/client/browser/templates/organisation.pt:104]: As mentioned the PR was developed when the first and last name were not a thing

File: src/euphorie/client/browser/organisation.py [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+        if not organisation_id:
+            return account.organisation
+
+        organisation = self.sqlsession.query(Organisation).get(
+            self.request.get(self.organisation_id_key)
+        )
+        if not organisation:
+            return
+
+        # Check if the user can manage users to the current organisation
```
[the remark this answers is not in the document]
Comment (ale-rt) [reply src/euphorie/client/browser/organisation.py:194]: > Why are the checks for manage permission done while retrieving the organisation record?

I think it is wise to protect the access to the data.
Somethink like `restrictedTraverse`, if you want.

> I would expect them to happen just before the actual change is performed.

This is totally not standard code.
If you want we can check the way you propose together.

File: src/euphorie/client/browser/organisation.py [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+            self.sqlsession.query(OrganisationMembership)
+            .filter(
+                sql.and_(
+                    OrganisationMembership.owner_id == organisation.owner_id,
+                    OrganisationMembership.member_id == account.id,
+                )
+            )
+            .count()
+        ):
+            # The user is a member, return the organisation
```
[the remark this answers is not in the document]
Comment (ale-rt) [reply src/euphorie/client/browser/organisation.py:350]: This is not a permission check, this is used to open the proper tab in the organisation view.

File: src/euphorie/client/browser/organisation.py [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+        if not organisation_id:
+            return account.organisation
+
+        organisation = self.sqlsession.query(Organisation).get(
+            self.request.get(self.organisation_id_key)
+        )
+        if not organisation:
+            return
+
+        # Check if the organisation should be visible to the current account
```
[the remark this answers is not in the document]
Comment (ale-rt) [reply src/euphorie/client/browser/organisation.py:460]: > The comment says “visible”, but the check is for the admin role. 

The comment was copy pasted, I fixed it.

> Also, I would expect this to happen just before the change is made (see other comment).

See my other reply ;)

File: src/euphorie/client/browser/organisation.py [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+    def membership(self):
+        """Return the role of the member in the organisation"""
+        membership_id = self.request.get("membership_id")
+        account = self.webhelpers.get_current_account()
+        membership = (
+            self.sqlsession.query(OrganisationMembership)
+            .filter(
+                OrganisationMembership.organisation_id == membership_id,
+                sql.or_(
+                    OrganisationMembership.owner_id == account.id,
```
[the remark this answers is not in the document]
Comment (ale-rt) [reply src/euphorie/client/browser/organisation.py:619]: See my other replies ;)

File: src/euphorie/client/browser/organisation.py [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
```
[the remark this answers is not in the document]
Comment (ale-rt) [reply src/euphorie/client/browser/organisation.py:692]: See my other reply ;)

**Comment by reinhardt** [comment]
I've slept on it and now I think the permission checks actually pretty closely resemble what is done in Plone (check permission on the view/content level). Maybe what irritated me was that everything happens in one sqlalchemy call chain. But I'm happy to merge like this after all.

The only thing that bothers me now is the `if not has_table("organisation"):` in the downgrade. I think it's supposed to be `if has_table("organisation"):`.

**Comment by ale-rt** [comment]
> The only thing that bothers me now is the `if not has_table("organisation"):` in the downgrade. I think it's supposed to be `if has_table("organisation"):`.

I already fixed that, haven't I?

**Comment by reinhardt** [comment]
Ah yes, that's 4cefbb3, thanks.

**Review by reinhardt — APPROVED** [review]

**Pull request merged**